R/dbpath_build.R

```diff
@@ -0,0 +1,98 @@
+#' Build a dbpath object
+#'
+#' Builds a [dbpath()] object from its parts. All parts are optional other than
+#' `dialect`. To create a database URL, use `format()` on the resulting `dbpath`
+#' object.
+#'
+#' @examples
+#' dbpath_build("postgres", host = "localhost", port = 5432, database = "money")
+#'
+#' @param dialect,driver The SQL database `dialect` and optional R `driver`
+#'   package to use with this dialect.
+#' @param username,password Optional username and password
+#' @param host,port Optional host and port
+#' @param database The name of the database
+#' @param params A named list of query parameters or a query string. For
+#'   example, either `list(foo = "bar", name = "value")` or
+#'   `foo=bar&name=value`.
+#'
+#' @return A [dbpath()] object.
+#'
+#' @export
+dbpath_build <- function(
+  dialect,
+  driver = NULL,
+  username = NULL,
+  password = NULL,
+  host = NULL,
+  port = NULL,
+  database = NULL,
+  params = NULL
+) {
+  # This function builds a single URL string following the dbpath format:
+  #
+  # <dialect>+<driver>://<username>:<password>@<host>:<port>/<database>?<params>
+  #
+  # It then hands it to dbpath() where it's parsed into a dbpath object. That's
+  # a bit circular, but it ensures that dbpath() and its parsing helpers remain
+  # the source of truth for the URL format and that we don't accidentally cut
+  # the wrong corners if we were to jump straight to a dbpath object.
+
+  assert_is_string(dialect)
+  url <- dialect
+
+  if (!is.null(driver)) {
+    assert_is_string(driver)
+    url <- paste0(url, "+", driver)
+  }
+
+  url <- paste0(url, "://")
+
+  if (!is.null(username)) {
+    assert_is_string(username)
+    url <- paste0(url, username)
+  }
+
+  if (!is.null(password)) {
+    assert_is_string(password)
+    url <- paste0(url, ":", password)
```

Comment:
This seems totally fine, but reminded me that we should maybe start URL encoding here, and decoding when we read the password part of a URL. (So passwords can contain reserved characters, like `@` or `:`)

Opened an issue to track / handle down the road https://github.com/machow/dbpath/issues/6 !

Reply:
Nice, I'm glad you called that out. I had the same thought and then decided I was already doing a lot in this PR. I think it'd be pretty straightforward though